File: chapter_03/python/3_7.py
S = "1234"
N = len(S)

sum = 0

# 2進数の1の部分に＋を入れるイメージ
for bit in range(1 << (N-1)):
    tmp = 0
    for i in range(N-1):
        tmp = tmp * 10 + int(S[i])

        # [+]を挿入する位置かどうかを確認
        if bit & (1 << i):
            sum += tmp
            tmp = 0

    # 最後の部分を加算
    tmp = tmp * 10 + int(S[-1])
    sum += tmp

print(f"sum : {sum}")

File: chapter_03/c++/3_7.cpp
#include <string>
#include <iostream>
#include <vector>

using namespace std;

int main(){
    string S = "125";

    int N = S.size();

    int sum = 0;

    // ２進数で１のところに＋を入れる
    for (int bit = 0; bit < (1 << N-1); bit++){
        int tmp = 0;
        for (int i = 0; i < N-1; i++){
            tmp = tmp * 10 + int(S[i] - '0');

            // +を挿入する位置かどうか確認
            if (bit & (1 << i)){
                sum += tmp;
                tmp = 0;
            }
        }

        //最後の部分を加算
        tmp = tmp * 10 + int(S[N - 1] - '0');
        sum += tmp;
    }

    cout << "sum : " << sum << endl;
}